From c50d13d70d3e499207589db21adb3db76c073c88 Mon Sep 17 00:00:00 2001
From: Poovendhan Selvaraj <quic_poovendh@quicinc.com>
Date: Wed, 19 Jul 2023 17:29:19 +0530
Subject: [PATCH 326/500] firmware: qcom_scm: enable secure crypto commands

secure crypto commands are used by dm-crypt driver for encryption
and decryption of data before being written into nand partition
and after being read from it.

enable these commands to enable secure nand feature.

Change-Id: Ia8934bd72e5b7268d84c8d06b151ef872e270157
Signed-off-by: Poovendhan Selvaraj <quic_poovendh@quicinc.com>
---
 drivers/crypto/qce/core.c              |  2 +-
 drivers/firmware/qcom_scm.c            | 18 ++++++++++++++++++
 drivers/firmware/qcom_scm.h            |  1 +
 include/linux/firmware/qcom/qcom_scm.h |  1 +
 4 files changed, 21 insertions(+), 1 deletion(-)

diff --git a/drivers/crypto/qce/core.c b/drivers/crypto/qce/core.c
index 9b33aeec7752..e737f08c39d5 100644
--- a/drivers/crypto/qce/core.c
+++ b/drivers/crypto/qce/core.c
@@ -266,7 +266,7 @@ static ssize_t fixed_sec_key_store(struct kobject *kobj,
 	if (use_fixed_key == 1) {
 		qce->use_fixed_key = true;
 	} else {
-		qti_qcekey_release_xpu_prot();
+		qcom_qcekey_release_xpu_prot();
 		qce->use_fixed_key = false;
 	}
 	return count;
diff --git a/drivers/firmware/qcom_scm.c b/drivers/firmware/qcom_scm.c
index 965adb61b27b..c35276a8c36d 100644
--- a/drivers/firmware/qcom_scm.c
+++ b/drivers/firmware/qcom_scm.c
@@ -400,6 +400,24 @@ int qcom_scm_set_cold_boot_addr(void *entry)
 }
 EXPORT_SYMBOL_GPL(qcom_scm_set_cold_boot_addr);
 
+/*
+ * qcom_qcekey_release_xpu_prot() - release XPU protection
+ */
+int qcom_qcekey_release_xpu_prot(void)
+{
+	int ret;
+	struct qcom_scm_res res;
+	struct qcom_scm_desc desc = {
+		.svc = QCOM_SCM_QCE_CRYPTO_SIP,
+		.cmd = QCOM_SCM_QCE_UNLOCK_CMD,
+		.arginfo = QCOM_SCM_ARGS(0, QCOM_SCM_VAL),
+	};
+
+	ret = qcom_scm_call(__scm->dev, &desc, &res);
+	return ret ? : res.result[0];
+}
+EXPORT_SYMBOL_GPL(qcom_qcekey_release_xpu_prot);
+
 /**
  * qcom_scm_cpu_power_down() - Power down the cpu
  * @flags:	Flags to flush cache
diff --git a/drivers/firmware/qcom_scm.h b/drivers/firmware/qcom_scm.h
index b08c55b783d7..c173e34955c4 100644
--- a/drivers/firmware/qcom_scm.h
+++ b/drivers/firmware/qcom_scm.h
@@ -168,6 +168,7 @@ extern int __qti_scm_tz_hvc_log(struct device *dev, u32 svc_id, u32 cmd_id,
 #define QCOM_SCM_QCE_CMD		0x3
 #define QCOM_SCM_QCE_CRYPTO_SIP		0xA
 #define QCOM_SCM_QCE_ENC_DEC_CMD	0xB
+#define QCOM_SCM_QCE_UNLOCK_CMD		0x4
 #define QCOM_SCM_SECCRYPT_CLRKEY_CMD	0xC
 extern int __qti_sec_crypt(struct device *dev, void *confBuf, int size);
 extern int __qti_seccrypt_clearkey(struct device *dev);
diff --git a/include/linux/firmware/qcom/qcom_scm.h b/include/linux/firmware/qcom/qcom_scm.h
index fca2a4c48cba..f3b2da894815 100644
--- a/include/linux/firmware/qcom/qcom_scm.h
+++ b/include/linux/firmware/qcom/qcom_scm.h
@@ -157,6 +157,7 @@ extern bool qcom_scm_lmh_dcvsh_available(void);
 extern int qti_seccrypt_clearkey(void);
 extern int qti_sec_crypt(void *buf, int size);
 extern int qti_set_qcekey_sec(void *buf, int size);
+extern int qcom_qcekey_release_xpu_prot(void);
 
 extern int qti_scm_is_tz_log_encrypted(void);
 extern int qti_scm_get_encrypted_tz_log(void *ker_buf, u32 buf_len, u32 log_id);
-- 
2.34.1

